calico-vpp-agent/prometheus/prometheus.go

```diff
@@ -38,236 +37,371 @@ import (
+	err := self.exporter.ExportMetric(
+		context.Background(),
+		nil, /* node */
+		nil, /* resource */
+		&metricspb.Metric{
+			MetricDescriptor: &metricspb.MetricDescriptor{
+				Name:        name,
+				Unit:        "",
+				Description: getVppTCPStatDescription(name),
+				Type:        metricspb.MetricDescriptor_CUMULATIVE_INT64,
+				LabelKeys: []*metricspb.LabelKey{
+					{Key: "worker", Description: "VPP worker index"},
 				},
 			},
+			Timeseries: []*metricspb.TimeSeries{{
+				LabelValues: []*metricspb.LabelValue{
+					{Value: strconv.Itoa(worker)},
+				},
+				Points: []*metricspb.Point{
+					{
+						Value: &metricspb.Point_Int64Value{
+							Int64Value: int64(value),
+						},
+					},
+				},
+			}},
 		},
+	)
+	if err != nil {
+		self.log.Errorf("Error prometheus exporter.ExportMetric for TCP %v", err)
 	}
 }
 
-func NewPrometheusServer(vpp *vpplink.VppLink, l *logrus.Entry) *Server {
-	server := &Server{
-		log:                      l,
-		vpp:                      vpp,
-		channel:                  make(chan common.CalicoVppEvent, 10),
-		podInterfacesByKey:       make(map[string]storage.LocalPodSpec),
-		podInterfacesBySwifIndex: make(map[uint32]storage.LocalPodSpec),
-	}
-	if *config.GetCalicoVppFeatureGates().PrometheusEnabled {
-		reg := common.RegisterHandler(server.channel, "prometheus events")
-		reg.ExpectEvents(common.PodAdded, common.PodDeleted)
+func (self *PrometheusServer) exportSessionMetric(name string, worker int, value uint64) {
+	err := self.exporter.ExportMetric(
+		context.Background(),
+		nil, /* node */
+		nil, /* resource */
+		&metricspb.Metric{
+			MetricDescriptor: &metricspb.MetricDescriptor{
+				Name:        cleanVppSessionStatName(name),
+				Unit:        "",
+				Description: getVppSessionStatDescription(name),
+				Type:        metricspb.MetricDescriptor_CUMULATIVE_INT64,
+				LabelKeys: []*metricspb.LabelKey{
+					{Key: "worker", Description: "VPP worker index"},
+				},
+			},
+			Timeseries: []*metricspb.TimeSeries{{
+				LabelValues: []*metricspb.LabelValue{
+					{Value: strconv.Itoa(worker)},
+				},
+				Points: []*metricspb.Point{
+					{
+						Value: &metricspb.Point_Int64Value{
+							Int64Value: int64(value),
+						},
+					},
+				},
+			}},
+		},
+	)
+	if err != nil {
+		self.log.Errorf("Error prometheus exporter.ExportMetric for Session %v", err)
 	}
-	return server
 }
 
-func (s *Server) ServePrometheus(t *tomb.Tomb) error {
+func (self *PrometheusServer) ServePrometheus(t *tomb.Tomb) error {
 	if !(*config.GetCalicoVppFeatureGates().PrometheusEnabled) {
 		return nil
 	}
-
-	s.log.Infof("Serve() Prometheus exporter")
+	self.log.Infof("Serve() Prometheus exporter")
 	go func() {
 		for t.Alive() {
 			/* Note: we will only receive events we ask for when registering the chan */
-			evt := <-s.channel
+			evt := <-self.channel
 			switch evt.Type {
 			case common.PodAdded:
+				self.lock.Lock()
+				defer self.lock.Unlock()
```

Comment:
`defer` will only run when you exit the function scope, so we either need to make the `case` statement a function or call `Unlock()` explicitly both in `continue` and at the end of the processing